crates/networking/p2p/sync.rs

```diff
@@ -114,48 +128,57 @@ impl SyncManager {
-                    .into_iter()
-                    .zip(all_block_hashes.into_iter())
-                {
-                    latest_block_number = header.number;
-                    store.set_canonical_block(header.number, hash)?;
-                    store.add_block_header(hash, header)?;
+                let mut pivot_idx = if all_block_headers.len() > MIN_FULL_BLOCKS {
+                    all_block_headers.len() - MIN_FULL_BLOCKS
+                } else {
+                    all_block_headers.len() - 1
+                };
+                let mut pivot_root = all_block_headers[pivot_idx].state_root;
+                let mut pivot_number = all_block_headers[pivot_idx].number;
+
+                let mut stale_pivot =
+                    !rebuild_state_trie(pivot_root, self.peers.clone(), store.clone()).await?;
+                // If the pivot became stale, set a further pivot and try again
+                if stale_pivot && pivot_idx != all_block_headers.len() - 1 {
+                    warn!("Stale pivot, switching to newer head");
+                    pivot_idx = all_block_headers.len() - 1;
+                    pivot_root = all_block_headers[pivot_idx].state_root;
+                    pivot_number = all_block_headers[pivot_idx].number;
+                    stale_pivot =
+                        !rebuild_state_trie(pivot_root, self.peers.clone(), store.clone()).await?;
                 }
-                // If all processes failed then they are likely to have a common cause (such as unaccessible storage), so return the first error
-                for result in set.join_all().await {
-                    result?;
+                if stale_pivot {
+                    warn!("Stale pivot, aborting sync");
+                    return Ok(());
                 }
-                // Set latest block number here to avoid reading state that is currently being synced
-                store.update_latest_block_number(latest_block_number)?;
+                // Wait for all bodies to be downloaded
+                let all_block_bodies = fetch_bodies_handle.await??;
+                // For all blocks before the pivot: Store the bodies and fetch the receipts
+                // For all blocks after the pivot: Process them fully
+                // let store_receipts_handle = tokio::spawn(store_receipts(all_block_hashes[pivot_index..]));
+                for (hash, (header, body)) in all_block_hashes.into_iter().zip(
+                    all_block_headers
+                        .into_iter()
+                        .zip(all_block_bodies.into_iter()),
+                ) {
+                    if header.number <= pivot_number {
+                        store.set_canonical_block(header.number, hash)?;
+                        store.add_block(Block::new(header, body))?;
+                    } else {
+                        store.set_canonical_block(header.number, hash)?;
+                        store.update_latest_block_number(header.number)?;
```

Comment:
idem, this is a job of the apply forkchoice update function, here we just store the block